dgl_sparse/src/softmax.cc

```diff
@@ -0,0 +1,97 @@
+/**
+ *  Copyright (c) 2022 by Contributors
+ * @file softmax.cc
+ * @brief DGL C++ Softmax operator implementation
+ */
+
+#include <sparse/sparse_matrix.h>
+#include <torch/script.h>
+
+#include "./matmul.h"
+#include "./utils.h"
+
+namespace dgl {
+namespace sparse {
+
+using namespace torch::autograd;
+
+class SoftmaxAutoGrad : public Function<SoftmaxAutoGrad> {
+ public:
+  static torch::Tensor forward(
+      AutogradContext* ctx, c10::intrusive_ptr<SparseMatrix> sparse_mat,
+      torch::Tensor sparse_val);
+
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+void _SoftmaxSanityCheck(
+    c10::intrusive_ptr<SparseMatrix> sparse_mat, torch::Tensor sparse_val) {
+  auto val_shape = sparse_val.sizes();
+  CHECK_EQ(val_shape[0], sparse_mat->nnz())
+      << "Softmax: the value shape does not match nnz of the sparse matrix.";
+  CHECK(sparse_val.device() == sparse_mat->device())
+      << "Softmax: sparse matrix and non-zero values should be on the same "
+         "device.";
+}
+
+torch::Tensor SoftmaxAutoGrad::forward(
+    AutogradContext* ctx, c10::intrusive_ptr<SparseMatrix> sparse_mat,
+    torch::Tensor sparse_val) {
+  auto sparse_val_max = SpMMNoAutoGrad(sparse_mat, sparse_val, "max");
+  auto sparse_val_exp =
+      SDDMMNoAutoGrad(sparse_mat, sparse_val, sparse_val_max, "sub").exp();
```

Comment:
Is this operation equal to a broadcast-sub between a matrix with a vector? Maybe instead of calling it `SDDMM`, we just call it `BroadcastSub`, `BroadcastDiv`, etc.

Reply:
Also, the grad operator of `BroadcastSub`/`BroadcastAdd` should be sparse reduction too, so supporting high-order grad shouldn't be too difficult too.

Reply:
Just FYI, this is introduced in https://github.com/dmlc/dgl/pull/5032. If rename, also think about rename SPMMNoAutoGrad.

Reply:
Renamed, autograd support can be left to future work